### PR TITLE
Team1 agent

### PR DESCRIPTION
**Pull request by mpardalos**
# Summary

This has been done entirely while pair programming with @E-Mac27.

Team 1's initial agent. Strategy:
* Foraging: 
  * forage deer with an amount inversely proportional to the sum of last turn's foraging
  * always share forage information
* Tax: always paid
* Request allocation when close to critical
* Steal from common pool when critical
* Gifts:
  * Request when critical
  * Opinion on other teams depends on how many gifts they gave us.
  * Give requested gifts if opinion of requester is positive, or if they are critical and opinion is neutral.

## Test Plan

Partial tests, see test files in team1 folder.

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

File: internal/clients/team1/client_test.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package team1
+
+import (
+//	"github.com/sajari/regression"
```
Comment (E-Mac27) [inline internal/clients/team1/client_test.go]: Are tests working with this commented out?

File: internal/clients/team1/iito.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+type opinionOnTeams struct {
+	clientID shared.ClientID
+	opinion  int
+}
+
+type sortByOpinion []opinionOnTeams
+
+/**************************/
+/*** 		Helpers	 	***/
+/**************************/
```
Comment (E-Mac27) [inline internal/clients/team1/iito.go:25]: D: nitpick

File: internal/clients/team1/iito.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	}
+	return requests
+}
+
+// GetGiftOffers allows clients to make offers in response to gift requests by other clients.
+// It can offer multiple partial gifts.
+// This will first iterate through a sorted array of opinionTeams starting from most favoured -> least favoured.
+// + If the opinion of that team is large enough, we just give resources and don't question it.
+// + If the opinion of that team is too low, we don't give resources :)
+// + If the opinion of that team is inbetween, we give resources if they are critical.
```
Comment (E-Mac27) [inline internal/clients/team1/iito.go]:
```suggestion
// + If the opinion of that team is in between, we give resources if they are critical.
```
linter :(

File: internal/clients/team1/iito.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+			} else if received <= 0 {
+				c.opinionTeams[i].opinion--
+			}
+		}
+	}
+}
+
+// DecideGiftAmount is executed at the end of each turn and asks clients how much
+// they want to fulfill a gift offer they have made.
+// Very similar to GetGiftOffers(). Only difference is the inbetween opinionated teams.
```
Comment (E-Mac27) [inline internal/clients/team1/iito.go]:
```suggestion
// Very similar to GetGiftOffers(). Only difference is the in between opinionated teams.
```
im sorry linter :(

File: internal/clients/team1/client.go [hunk 5]
```diff
@@ -2,16 +2,167 @@
...
+	Anxious
+	Desperate
+)
+
+func (st EmotionalState) String() string {
+	strings := [...]string{"Normal", "Anxious", "Desperate"}
+	if st >= 0 && int(st) < len(strings) {
+		return strings[st]
+	}
+	return fmt.Sprintf("UNKNOWN ForageType '%v'", int(st))
```
Comment (lhl2617) [inline internal/clients/team1/client.go]:
```suggestion
	return fmt.Sprintf("UNKNOWN EmotionalState '%v'", int(st))
```

File: internal/server/server.go [hunk 6]
```diff
@@ -34,8 +36,15 @@ type SOMASServer struct {
 
 // NewSOMASServer returns an instance of the main server we use.
 func NewSOMASServer(gameConfig config.Config) Server {
+	clients := map[shared.ClientID]baseclient.Client{}
+	for _, id := range shared.TeamIDs[:2] {
+		clients[id] = team1.NewClient(id)
+	}
+	for _, id := range shared.TeamIDs[2:] {
+		clients[id] = baseclient.RegisteredClients[id]
+	}
```
Comment (lhl2617) [inline internal/server/server.go]: these shouldn't be here?

File: internal/common/baseclient/basepresident.go [hunk 7]
```diff
@@ -58,7 +58,7 @@ func (p *BasePresident) SetTaxationAmount(islandsResources map[shared.ClientID]s
 
 	for clientID, clientReport := range islandsResources {
 		if clientReport.Reported {
-			taxAmountMap[clientID] = shared.Resources(float64(clientReport.ReportedAmount) * rand.Float64())
+			taxAmountMap[clientID] = shared.Resources(float64(clientReport.ReportedAmount) * (rand.Float64() * 0.1 + 0.1))
```
Comment (lhl2617) [inline internal/common/baseclient/basepresident.go]: base client changes should be in separate PR?

File: internal/clients/team1/iito.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+)
+
+/**************************/
+/*** 		Types	 	***/
+/**************************/
+
+// optionOnTeamsDict contains the opinion of Team1 about other teams.
+// 0 is neutral, Positive -> Positive Opinion, Negative -> Negative Opinion
+type opinionOnTeams struct {
```
Comment (lhl2617) [inline internal/clients/team1/iito.go]: should be `opinionOnTeam` I think (only one team)

File: internal/clients/team1/iito.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	"sort"
+
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+)
+
+/**************************/
+/*** 		Types	 	***/
+/**************************/
+
+// optionOnTeamsDict contains the opinion of Team1 about other teams.
```
Comment (lhl2617) [inline internal/clients/team1/iito.go]:
```suggestion
// optionOnTeam contains the opinion of Team1 about other teams.
```

File: internal/clients/team1/foraging.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+	r := new(regression.Regression)
+	r.SetObserved("Team1 reward")
+	r.SetVar(0, "Team1 contribution")
+
+	for _, outcome := range history {
+		r.Train(regression.DataPoint(float64(outcome.revenue), []float64{float64(outcome.contribution)}))
+	}
+
+	r.AddCross(regression.PowCross(0, 2))
+	r.Run()
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go]: `Run` returns error... is this fine to ignore?

File: internal/clients/team1/foraging.go [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+	bestReward := shared.Resources(0)
+	var decision shared.ForageDecision
+
+	for forageType := range c.forageHistory {
+		r := outcomeRegression(c.forageHistory[forageType])
+		contribution := c.regressionOptimalContribution(r)
+		if contribution > c.gameState().ClientInfo.Resources {
+			contribution = c.gameState().ClientInfo.Resources
+		}
+		expectedRewardF, _ := r.Predict([]float64{float64(contribution)})
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go]: ignored error

File: internal/clients/team1/foraging.go [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+				Type:         forageType,
+				Contribution: contribution,
+			}
+		}
+	}
+
+	return decision
+}
+
+const flipScale = 0.3
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go]: meaning?

File: internal/clients/team1/foraging.go [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+/************************/
+/***  Forage Helpers  ***/
+/************************/
+
+// bestROIForageType finds the ForageType that has the best average ROI. If all
+// forageTypes have negative returns then it returns shared.ForageType(-1)
+func bestROIForageType(forageHistory ForageHistory) shared.ForageType {
+	bestForageType := shared.ForageType(-1)
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go:241]: not a fan of this, let's open an issue/pr for a new ForageType called `NoForage`?

or, change the return signature of the func to `(shared.ForageType, participating bool)`

Comment (lhl2617) [reply internal/clients/team1/foraging.go:241]: (lowpri nitpick)

File: internal/clients/team1/foraging.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+		// Big contribution
+		return shared.ForageDecision{
+			Contribution: 0.3 * c.gameState().ClientInfo.Resources,
+			Type:         shared.DeerForageType,
+		}
+	}
+
+	// Proxy for population
+	totalROI := totalRevenueLastTurn / totalContributionLastTurn
+	averageContribution := totalContributionLastTurn / shared.Resources(totalHuntersLastTurn)
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go:140]: ~vibe~ div0 check

File: go.mod [hunk 15]
```diff
@@ -5,6 +5,8 @@ go 1.15
 require (
 	github.com/alessio/shellescape v1.4.1 // indirect
 	github.com/pkg/errors v0.9.1
+	// Team 1 agent
```
Comment (lhl2617) [inline go.mod]: fine, you can delete :)

File: internal/clients/team1/client.go [hunk 16]
```diff
@@ -2,16 +2,167 @@
...
+	// desperateStealAmount is the max randomness we will add to our foraging
+	// amount as a percentage of current resources
+	forageContributionNoisePercent float64
+
+	// maxOpinion is the boundary where we either give resources without questioning
+	// or we refuse to give them resources.
+	maxOpinion int
 }
 
+// client is Lucy.
```
Comment (lhl2617) [inline internal/clients/team1/client.go:68]: nice

File: internal/clients/team1/client_test.go [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+	"math"
+	"testing"
+
+	"github.com/SOMAS2020/SOMAS2020/internal/common/gamestate"
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+)
+
+type testServerHandle struct {
+	clientGameState gamestate.ClientGameState
```
Comment (lhl2617) [inline internal/clients/team1/client_test.go]: i believe you can silence the currently failing test if you add `server.ServerForClient` in here.

File: internal/clients/team1/client_test.go (hunk 1 above)
Comment (lhl2617) [reply internal/clients/team1/client_test.go]: they should be, it's just goimport is really pedantic about the order of imports

File: internal/clients/team1/foraging.go [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+func (c *client) regressionOptimalContribution(r regression.Regression) shared.Resources {
+	if r.Coeff(2) > 0 {
+		// Curves up, mo money is mo money
+		return 0.4 * c.gameState().ClientInfo.Resources
+	}
+
+	// Curves down, mo money is not always mo money
+	// c.Logf("%v x^2 + %v x + %v", r.Coeff(2), r.Coeff(1), r.Coeff(0))
+	// c.Logf("%v / %v = %v ", -r.Coeff(1), 2*r.Coeff(2), shared.Resources(-r.Coeff(1)/(2*r.Coeff(2))))
+	return shared.Resources(-r.Coeff(1) / (2 * r.Coeff(2)))
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go:185]: ~vibe~ div0 check

File: internal/clients/team1/foraging.go [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+
+	for forageType, outcomes := range forageHistory {
+		ROIsum := 0.0
+		for _, outcome := range outcomes {
+			if outcome.contribution != 0 {
+				ROIsum += outcome.ROI()
+			}
+		}
+
+		averageROI := ROIsum / float64(len(outcomes))
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go:256]: ~vibe~ div0 check -- unless it's guaranteed (i assume so otherwise this wont even run)

File: internal/server/server.go (hunk 6 above)
Comment (mpardalos) [reply internal/server/server.go]: indeed, left over from testing

File: internal/common/baseclient/basepresident.go (hunk 7 above)
Comment (mpardalos) [reply internal/common/baseclient/basepresident.go]: also indeed, moving to another PR

File: internal/clients/team1/foraging.go (hunk 12 above)
Comment (mpardalos) [reply internal/clients/team1/foraging.go]: It's a scale factor for the contribution of flipForage. It should actually go to the agent config.

File: internal/clients/team1/foraging.go (hunk 13 above)
Comment (mpardalos) [reply internal/clients/team1/foraging.go:241]: I've already suggested this in #143, but JamesTev said that not foraging should just be 0 resources.

File: internal/clients/team1/foraging.go (hunk 19 above)
Comment (mpardalos) [reply internal/clients/team1/foraging.go:256]: actually it would if forageHistory has an empty (not missing, empty) history for a certain forageType. That should basically never happen, so I'll make it a panic.

File: internal/clients/team1/iito.go (hunk 2 above)
Comment (mpardalos) [reply internal/clients/team1/iito.go:25]: That's weird, it looks aligned on my computer.

**Review by lhl2617 — APPROVED** [review]
- [x] all changes only in client folder
- [x] no panic or breaky stuff

gogogo

File: internal/clients/team1/foraging.go (hunk 13 above)
Comment (mpardalos) [reply internal/clients/team1/foraging.go:241]: Should I just leave it as-is?

**Comment by mpardalos** [comment]
DO NOT MERGE. It bronk after changes to main

File: internal/clients/team1/iito.go (hunk 2 above)
Comment (lhl2617) [reply internal/clients/team1/iito.go:25]: emacs reeeeee

**Comment by mpardalos** [comment]
It's fine now. Merging.
